modules/diskq/qdisk.c

```diff
@@ -569,18 +555,47 @@ qdisk_remove_head(QDisk *self)
   return TRUE;
 }
 
+static gboolean
+_overwrite_with_real_record_length(GString *serialized)
+{
+  guint32 record_length = GUINT32_TO_BE(serialized->len - sizeof(guint32));
+  if (record_length == 0)
+    return FALSE;
+
+  g_string_overwrite_len(serialized, 0, (gchar *) &record_length, sizeof(guint32));
+  return TRUE;
+}
+
 gboolean
 qdisk_serialize_msg(QDisk *self, LogMessage *msg, GString *serialized)
 {
   gboolean result = TRUE;
   SerializeArchive *sa = serialize_string_archive_new(serialized);
 
+  /* Leave space for the real record_length for later */
+  if (!serialize_write_uint32(sa, 0))
+    {
+      result = FALSE;
+      goto exit;
+    }
+
   if (!log_msg_serialize(msg, sa, self->options->compaction ? LMSF_COMPACTION : 0))
     {
-      g_string_truncate(serialized, 0);
       result = FALSE;
+      goto exit;
     }
 
+  if (!_overwrite_with_real_record_length(serialized))
+    {
+      msg_error("Error serializing empty message for the disk-queue file");
```

Comment:
Please provide identifiers for investigating later on (like diskq filename).

Reply:
I added the filename to the error message, and also modified the code, to report errors from other failure points of this function.